File: packages/icoder-sdk/src/resources/runtime.ts
/** Runtime API resource — agent execution, lifecycle, observability. */

import type { AxiosInstance } from 'axios';
import { requestConfig, type iCoDerRequestOptions } from '../request-options.js';

export interface InstalledAgent {
  agent_ref: string; name: string; version: string; description: string;
  category: string; icon: string; agent_type: string; status: string;
  installed_at: string; publisher_name: string;
  expert_count: number; tool_count: number; tier: number; usage_count: number;
}

export interface RuntimeRunResult {
  run_id: string; agent_ref: string; status: string; output: string;
  primary_diagnosis: Record<string, unknown>;
  secondary_diagnoses: Record<string, unknown>[];
  procedures: Record<string, unknown>[];
  issues_found: Record<string, unknown>[];
  audit_trail: Record<string, unknown>[];
  processing_time_ms: number;
  safety?: Record<string, unknown>;
}

export class RuntimeResource {
  constructor(private http: AxiosInstance) {}

  status(options?: iCoDerRequestOptions) {
    return this.http.get('/api/runtime/status', requestConfig(options));
  }
  dataPolicy(options?: iCoDerRequestOptions) {
    return this.http.get('/api/runtime/data-policy', requestConfig(options));
  }

  listAgents(agentType = '', options?: iCoDerRequestOptions) {
    return this.http.get<{ agents: InstalledAgent[]; total: number }>(
      '/api/runtime/agents', requestConfig(options, { agent_type: agentType }),
    );
  }

  installAgent(
    name: string,
    version: string,
    agentType = 'community',
    options?: iCoDerRequestOptions,
  ) {
    return this.http.post(
      '/api/runtime/agents/install',
      { agent_name: name, agent_version: version, agent_type: agentType },
      requestConfig(options),
    );
  }

  runAgent(agentRef: string, input: string, options?: iCoDerRequestOptions) {
    return this.http.post<RuntimeRunResult>(
      `/api/runtime/agents/${encodeURIComponent(agentRef)}/run`,
      { input },
      requestConfig(options),
    );
  }

  agentLifecycle(
    agentRef: string,
    action: 'enable' | 'disable' | 'uninstall' | 'rollback',
    options?: iCoDerRequestOptions,
  ) {
    return this.http.post(
      `/api/runtime/agents/${encodeURIComponent(agentRef)}/lifecycle`,
      { action },
      requestConfig(options),
    );
  }

  listRuns(agentRef = '', limit = 50, options?: iCoDerRequestOptions) {
    return this.http.get(
      '/api/runtime/runs', requestConfig(options, { agent_ref: agentRef, limit }),
    );
  }
  getRun(runId: string, options?: iCoDerRequestOptions) {
    return this.http.get(
      `/api/runtime/runs/${encodeURIComponent(runId)}`, requestConfig(options),
    );
  }
  fallbackStats(hours = 24, options?: iCoDerRequestOptions) {
    return this.http.get(
      '/api/runtime/observability/fallback', requestConfig(options, { hours }),
    );
  }
  shadowStats(hours = 24, options?: iCoDerRequestOptions) {
    return this.http.get(
      '/api/runtime/observability/shadow', requestConfig(options, { hours }),
    );
  }
  auditLog(eventType = '', limit = 100, options?: iCoDerRequestOptions) {
    return this.http.get(
      '/api/runtime/audit-log',
      requestConfig(options, { event_type: eventType, limit }),
    );
  }
  medicalCodingStatus(options?: iCoDerRequestOptions) {
    return this.http.get('/api/runtime/medical-coding/status', requestConfig(options));
  }
  testMedicalCoding(text: string, options?: iCoDerRequestOptions) {
    return this.http.post(
      '/api/runtime/medical-coding/test',
      { encounter_text: text },
      requestConfig(options),
    );
  }
  ruleEngineStatus(options?: iCoDerRequestOptions) {
    return this.http.get('/api/runtime/rule-engine/status', requestConfig(options));
  }
  ruleEngineRules(options?: iCoDerRequestOptions) {
    return this.http.get('/api/runtime/rule-engine/rules', requestConfig(options));
  }
  validateRules(
    ruleSet: string,
    output: Record<string, unknown>,
    context: Record<string, unknown> = {},
    options?: iCoDerRequestOptions,
  ) {
    return this.http.post(
      '/api/runtime/rule-engine/validate',
      { rule_set: ruleSet, structured_output: output, context },
      requestConfig(options),
    );
  }
  registryHealth(options?: iCoDerRequestOptions) {
    return this.http.get('/api/runtime/registry/health', requestConfig(options));
  }
  registryRepair(direction = 'registry_to_db', options?: iCoDerRequestOptions) {
    return this.http.post(
      '/api/runtime/registry/repair', { direction }, requestConfig(options),
    );
  }
}

File: packages/icoder-python/icoder_sdk/resources/runtime.py
"""Legacy runtime lifecycle and observability resources backed by real routes."""

from __future__ import annotations

from typing import Any, Literal
from urllib.parse import quote

from ..client import iCoDerClient
from ..request_options import RequestOptions


RuntimeLifecycleAction = Literal["enable", "disable", "uninstall", "rollback"]


def _bounded_integer(value: int, name: str, *, minimum: int, maximum: int) -> int:
    if (
        not isinstance(value, int)
        or isinstance(value, bool)
        or value < minimum
        or value > maximum
    ):
        raise ValueError(f"{name} must be an integer between {minimum} and {maximum}")
    return value


class RuntimeResource:
    def __init__(self, client: iCoDerClient):
        self._client = client

    def _json(self, method: str, path: str, *, request_options=None, **kwargs):
        response = self._client.request(
            method,
            path,
            request_options=request_options,
            **kwargs,
        )
        response.raise_for_status()
        return response.json()

    def status(self, request_options: RequestOptions | None = None) -> dict[str, Any]:
        return self._json("GET", "/api/runtime/status", request_options=request_options)

    def data_policy(
        self, request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "GET", "/api/runtime/data-policy", request_options=request_options,
        )

    def list_agents(
        self,
        agent_type: str = "",
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "GET",
            "/api/runtime/agents",
            params={"agent_type": agent_type},
            request_options=request_options,
        )

    def install_agent(
        self,
        name: str,
        version: str,
        agent_type: str = "community",
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "POST",
            "/api/runtime/agents/install",
            json={
                "agent_name": name,
                "agent_version": version,
                "agent_type": agent_type,
            },
            request_options=request_options,
        )

    def run_agent(
        self,
        agent_ref: str,
        input_text: str,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "POST",
            f"/api/runtime/agents/{quote(agent_ref, safe='')}/run",
            json={"input": input_text},
            request_options=request_options,
        )

    def agent_lifecycle(
        self,
        agent_ref: str,
        action: RuntimeLifecycleAction,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        if action not in {"enable", "disable", "uninstall", "rollback"}:
            raise ValueError("action must be enable, disable, uninstall, or rollback")
        return self._json(
            "POST",
            f"/api/runtime/agents/{quote(agent_ref, safe='')}/lifecycle",
            json={"action": action},
            request_options=request_options,
        )

    def list_runs(
        self,
        agent_ref: str = "",
        limit: int = 50,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        _bounded_integer(limit, "limit", minimum=1, maximum=200)
        return self._json(
            "GET",
            "/api/runtime/runs",
            params={"agent_ref": agent_ref, "limit": limit},
            request_options=request_options,
        )

    def get_run(
        self,
        run_id: str,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "GET",
            f"/api/runtime/runs/{quote(run_id, safe='')}",
            request_options=request_options,
        )

    def fallback_stats(
        self,
        hours: int = 24,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        _bounded_integer(hours, "hours", minimum=1, maximum=168)
        return self._json(
            "GET",
            "/api/runtime/observability/fallback",
            params={"hours": hours},
            request_options=request_options,
        )

    def shadow_stats(
        self,
        hours: int = 24,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        _bounded_integer(hours, "hours", minimum=1, maximum=168)
        return self._json(
            "GET",
            "/api/runtime/observability/shadow",
            params={"hours": hours},
            request_options=request_options,
        )

    def audit_log(
        self,
        event_type: str = "",
        limit: int = 100,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        _bounded_integer(limit, "limit", minimum=1, maximum=500)
        return self._json(
            "GET",
            "/api/runtime/audit-log",
            params={"event_type": event_type, "limit": limit},
            request_options=request_options,
        )

    def medical_coding_status(
        self, request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "GET",
            "/api/runtime/medical-coding/status",
            request_options=request_options,
        )

    def test_medical_coding(
        self,
        text: str,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "POST",
            "/api/runtime/medical-coding/test",
            json={"encounter_text": text},
            request_options=request_options,
        )

    def rule_engine_status(
        self, request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "GET", "/api/runtime/rule-engine/status", request_options=request_options,
        )

    def rule_engine_rules(
        self, request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "GET", "/api/runtime/rule-engine/rules", request_options=request_options,
        )

    def validate_rules(
        self,
        rule_set: str,
        output: dict[str, Any],
        context: dict[str, Any] | None = None,
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "POST",
            "/api/runtime/rule-engine/validate",
            json={
                "rule_set": rule_set,
                "structured_output": output,
                "context": dict(context or {}),
            },
            request_options=request_options,
        )

    def registry_health(
        self, request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "GET", "/api/runtime/registry/health", request_options=request_options,
        )

    def registry_repair(
        self,
        direction: str = "registry_to_db",
        request_options: RequestOptions | None = None,
    ) -> dict[str, Any]:
        return self._json(
            "POST",
            "/api/runtime/registry/repair",
            json={"direction": direction},
            request_options=request_options,
        )
